### ISSUE TITLE
DateTimeInterface&!\DateTime should be understood to \DateTimeImmutable

### ISSUE THREAD

**Issue by VincentLanglet** [vimeo/psalm#6152]
Even if it's purely true, because `DateTimeInterface::add` is not in the interface, since only \DateTime and \DatetimeImutable can implement the interface, all the common method should be added to the interface IMHO (with a stub ?).

For instance 
```
\assert($date instanceof \DateTime || $date instanceof \DateTimeImmutable)
```
is now considered as always true by phpstan, and maybe one day psalm will considered as always true too.

So I don't expect such
```
UndefinedInterfaceMethod - Method DateTimeInterface::add does not exist 
```
error.

**Comment by VincentLanglet** [comment]
This is a snippet https://psalm.dev/r/030cd55cfb

At least https://psalm.dev/r/07ed0b7503 is working fine with phpstan https://phpstan.org/r/9c1b8f91-7322-4c12-bc32-6b77a39b4dd5

This is essentially because 
```
\assert($date instanceof \DateTime || $date instanceof \DateTimeImmutable)
```
is always true for phpstan.

At least changing this for psalm would be useful.
Any idea how to do it @orklah ?

**Comment by weirdan** [comment]
That would be horribly unsafe. If you're calling `add()` on a `DateTimeInterface` you may, or may not change the instance you had received. That's the reason why `DateTimeInterface` doesn't have any mutation methods.

BTW, PHPStan rejects that too: https://phpstan.org/r/00256fe8-1c98-4fdd-acc7-0724ad1304ca

**Comment by VincentLanglet** [comment]
> That would be horribly unsafe. If you're calling `add()` on a `DateTimeInterface` you may, or may not change the instance you had received. That's the reason why `DateTimeInterface` doesn't have any mutation methods.
> 
> BTW, PHPStan rejects that too: https://phpstan.org/r/00256fe8-1c98-4fdd-acc7-0724ad1304ca

I understand that point.

Can we re-open the issue and update it (or should I create another) to at least solve those issue @weirdan ?
https://psalm.dev/r/07ed0b7503
https://psalm.dev/r/105687a253

**Comment by weirdan** [comment]
With the updated title it makes sense. There's a small number of widely used interfaces in PHP that have a closed set of descendants (e.g. `Throwable`, `Traversable`) that may benefit from being treated similarly as well.

**Comment by VincentLanglet** [comment]
I can try a PR with some help about where I should look at in the code ; I may be wrong but I suppose this is not an hard thing to change

**Comment by weirdan** [comment]
You can start by running Psalm with `--debug-emitted-issues` flag, on a sample code, and then following the backtrace it emits.

**Comment by someniatko** [comment]
Just a random thought: similar behavior could be applied to enums, I suppose.

**Comment by ondrejmirtes** [comment]
@someniatko Can you elaborate on that?

**Comment by someniatko** [comment]
@ondrejmirtes Given that enums have a limited set of possible values, similar inferring could be made, e.g.

```php
enum Parity { case Odd; case Even; }

function getParity(): Parity { ... }

$parity = getParity();
if ($parity !== Parity::Odd) {
  // infer that $parity is Parity::Even
}
```

**Comment by Jack97** [comment]
I'll pick this issue up, pretty sure it will only involve a small change to the `NegatedAssertionReconciler`.

**Comment by Jack97** [comment]
This issue can be closed now, it was fixed in #6181

**Comment by fox91** [comment]
For reference only: https://bugs.php.net/bug.php?id=75195
